File: e2e/dashboard.spec.ts
import { test, expect } from "./fixtures/url-tracker";
import { waitForPageReady } from "./helpers/screenshot-helper";
import { signIn, ADMIN_USER } from "./helpers/auth-helper";

test.describe("User Dashboard", () => {
  test.beforeEach(async ({ page }) => {
    await signIn(page, ADMIN_USER.email, ADMIN_USER.password);
  });

  test("shows stat cards", async ({ page }) => {
    await expect(page.locator(".stat-card")).toHaveCount(4);
    await expect(page.locator("body")).toContainText("Jobs Found");
    await expect(page.locator("body")).toContainText("High Matches");
    await expect(page.locator("body")).toContainText("Applied");
    await expect(page.locator("body")).toContainText("Needs Attention");
  });

  test("shows recent job listings section", async ({ page }) => {
    await expect(page.locator("body")).toContainText("Recent Job Listings");
  });

  test("shows recent activity section", async ({ page }) => {
    await expect(page.locator("body")).toContainText("Recent Activity");
  });

  test("sidebar navigation links are visible", async ({ page }) => {
    await expect(page.getByRole("link", { name: /Dashboard/ })).toBeVisible();
    await expect(page.getByRole("link", { name: /Job Sources/ })).toBeVisible();
    await expect(page.getByRole("link", { name: /Job Listings/ })).toBeVisible();
    await expect(page.getByRole("link", { name: /Applications/ })).toBeVisible();
    await expect(page.getByRole("link", { name: /Analytics/ })).toBeVisible();
    await expect(page.getByRole("link", { name: /Profile/ })).toBeVisible();
    await expect(page.getByRole("link", { name: /Settings/ })).toBeVisible();
  });

  test("dark mode toggle works", async ({ page }) => {
    const toggleBtn = page.locator("button", { hasText: /🌙|☀️/ });
    await expect(toggleBtn).toBeVisible();
    await toggleBtn.click();
    await expect(page.locator("html")).toHaveAttribute("data-theme", "dark");
    await toggleBtn.click();
    await expect(page.locator("html")).toHaveAttribute("data-theme", "light");
  });
});

File: spec/requests/dashboard_spec.rb
require "rails_helper"

RSpec.describe "Dashboard", type: :request do
  let(:user) { create(:user) }

  before { sign_in user }

  describe "GET /dashboard" do
    it "renders the dashboard page" do
      get dashboard_path
      expect(response).to have_http_status(:success)
      expect(response.body).to include("Dashboard")
    end

    it "shows stat cards" do
      get dashboard_path
      expect(response.body).to include("Jobs Found")
      expect(response.body).to include("High Matches")
      expect(response.body).to include("Applied")
      expect(response.body).to include("Needs Attention")
    end

    it "shows recent job listings section" do
      get dashboard_path
      expect(response.body).to include("Recent Job Listings")
    end

    it "shows recent activity section" do
      get dashboard_path
      expect(response.body).to include("Recent Activity")
    end

    it "redirects unauthenticated users to sign in" do
      sign_out user
      get dashboard_path
      expect(response).to redirect_to(new_user_session_path)
    end
  end
end
